File: src/tracked_hardware/tracked_hardware/esp32_bridge.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Imu, NavSatFix, NavSatStatus
import serial
import struct
import threading

FRAME_HEADER1  = 0xAA
FRAME_HEADER2  = 0x55
PKT_ID_CMD_VEL = 0x01
PKT_ID_IMU     = 0x10
PKT_ID_GPS     = 0x11  # ESP32 GPS Paket ID

class ESP32Bridge(Node):
    def __init__(self):
        super().__init__('esp32_bridge')

        self.declare_parameter('port', '/dev/ttyAMA0')
        self.declare_parameter('baudrate', 115200)
        self.declare_parameter('track_width', 0.22)
        self.declare_parameter('slip_factor', 1.25)
        self.declare_parameter('imu_frame_id', 'imu_link')
        self.declare_parameter('gps_frame_id', 'gps_link')

        port = self.get_parameter('port').value
        baud = self.get_parameter('baudrate').value
        self.effective_width = self.get_parameter('track_width').value * self.get_parameter('slip_factor').value
        self.imu_frame_id = self.get_parameter('imu_frame_id').value
        self.gps_frame_id = self.get_parameter('gps_frame_id').value

        try:
            self.ser = serial.Serial(port, baud, timeout=0.1)
            self.get_logger().info(f"ESP32 UART acildi: {port} @ {baud}")
        except Exception as e:
            self.get_logger().fatal(f"Seri port acilamadi: {e}")
            raise e

        # ROS 2 Arayüzleri
        self.sub_cmd_vel = self.create_subscription(Twist, '/cmd_vel', self.cmd_vel_cb, 10)
        self.pub_imu = self.create_publisher(Imu, '/imu/data_raw', 10)
        self.pub_gps = self.create_publisher(NavSatFix, '/gps/fix', 10)

        self.running = True
        self.rx_thread = threading.Thread(target=self.rx_loop, daemon=True)
        self.rx_thread.start()

    def calc_crc(self, pkt_id, length, payload):
        crc = pkt_id ^ length
        for b in payload:
            crc ^= b
        return crc

    def cmd_vel_cb(self, msg: Twist):
        vx = msg.linear.x
        wz = msg.angular.z
        v_left = vx - (wz * self.effective_width / 2.0)
        v_right = vx + (wz * self.effective_width / 2.0)

        payload = struct.pack('<ff', float(v_left), float(v_right))
        length = len(payload)
        crc = self.calc_crc(PKT_ID_CMD_VEL, length, payload)

        packet = bytearray([FRAME_HEADER1, FRAME_HEADER2, PKT_ID_CMD_VEL, length]) + bytearray(payload) + bytearray([crc])
        try:
            self.ser.write(packet)
        except Exception as e:
            self.get_logger().warn(f"TX Hatasi: {e}")

    def rx_loop(self):
        state = 'H1'
        pkt_id, pkt_len = 0, 0
        payload = bytearray()

        while self.running and rclpy.ok():
            try:
                byte_in = self.ser.read(1)
                if not byte_in:
                    continue
                b = byte_in[0]

                if state == 'H1':
                    if b == FRAME_HEADER1: state = 'H2'
                elif state == 'H2':
                    state = 'ID' if b == FRAME_HEADER2 else 'H1'
                elif state == 'ID':
                    pkt_id = b
                    state = 'LEN'
                elif state == 'LEN':
                    pkt_len = b
                    payload.clear()
                    state = 'PAYLOAD' if pkt_len > 0 else 'CRC'
                elif state == 'PAYLOAD':
                    payload.append(b)
                    if len(payload) >= pkt_len:
                        state = 'CRC'
                elif state == 'CRC':
                    if self.calc_crc(pkt_id, pkt_len, payload) == b:
                        self.process_packet(pkt_id, payload)
                    state = 'H1'
            except Exception:
                state = 'H1'

    def process_packet(self, pkt_id, payload):
        stamp = self.get_clock().now().to_msg()

        # IMU (24 Bayt: 6x float32)
        if pkt_id == PKT_ID_IMU and len(payload) == 24:
            ax, ay, az, gx, gy, gz = struct.unpack('<ffffff', payload)

            msg = Imu()
            msg.header.stamp = stamp
            msg.header.frame_id = self.imu_frame_id
            msg.linear_acceleration.x = ax
            msg.linear_acceleration.y = ay
            msg.linear_acceleration.z = az
            msg.angular_velocity.x = gx
            msg.angular_velocity.y = gy
            msg.angular_velocity.z = gz
            msg.orientation_covariance[0] = -1.0
            self.pub_imu.publish(msg)

        # GPS (17 Bayt: double lat, double lon, uint8 fix)
        elif pkt_id == PKT_ID_GPS and len(payload) == 17:
            lat, lon, fix = struct.unpack('<ddB', payload)

            gps_msg = NavSatFix()
            gps_msg.header.stamp = stamp
            gps_msg.header.frame_id = self.gps_frame_id
            gps_msg.status.status = NavSatStatus.STATUS_FIX if fix > 0 else NavSatStatus.STATUS_NO_FIX
            gps_msg.status.service = NavSatStatus.SERVICE_GPS
            gps_msg.latitude = lat
            gps_msg.longitude = lon
            gps_msg.altitude = 0.0
            self.pub_gps.publish(gps_msg)

    def destroy_node(self):
        self.running = False
        if hasattr(self, 'ser') and self.ser.is_open:
            self.ser.close()
        super().destroy_node()

def main(args=None):
    rclpy.init(args=args)
    node = ESP32Bridge()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/tracked_hardware/src/esp32_bridge.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <sensor_msgs/msg/nav_sat_fix.hpp>
#include <sensor_msgs/msg/nav_sat_status.hpp>
#include <sensor_msgs/msg/temperature.hpp>
#include "geometry_msgs/msg/twist_with_covariance_stamped.hpp"

#include <thread>
#include <atomic>
#include <algorithm>
#include <vector>
#include <cstring>

#include "tracked_hardware/protocol.hpp"
#include "tracked_hardware/uart_driver.hpp"

class ESP32Bridge : public rclcpp::Node {
public:
    ESP32Bridge() : Node("esp32_bridge"), running_(false) {
        // PARAMETERS: Robot kinematics and serial communication options
        this->declare_parameter<std::string>("port", "/dev/ttyAMA0");
        this->declare_parameter<int>("baudrate", 115200);
        this->declare_parameter<double>("track_width", 0.22);
        this->declare_parameter<double>("slip_factor", 1.25);
        this->declare_parameter<std::string>("imu_frame_id", "imu_link");
        this->declare_parameter<std::string>("gps_frame_id", "gps_link");

        std::string port = this->get_parameter("port").as_string();
        int baudrate = this->get_parameter("baudrate").as_int();
        double track_width = this->get_parameter("track_width").as_double();
        double slip_factor = this->get_parameter("slip_factor").as_double();

        effective_width_ = track_width * slip_factor;
        imu_frame_id_ = this->get_parameter("imu_frame_id").as_string();
        gps_frame_id_ = this->get_parameter("gps_frame_id").as_string();

        if (!uart_.open_port(port, baudrate)) {
            RCLCPP_FATAL(this->get_logger(), "Failed to open UART port: %s", port.c_str());
            throw std::runtime_error("UART connection error");
        }
        RCLCPP_INFO(this->get_logger(), "UART initialized: %s @ %d baud", port.c_str(), baudrate);

        // ROS: Publishers and subscriptions
        sub_cmd_vel_ = this->create_subscription<geometry_msgs::msg::Twist>(
            "/cmd_vel", 10, std::bind(&ESP32Bridge::cmd_vel_callback, this, std::placeholders::_1));
        pub_imu_ = this->create_publisher<sensor_msgs::msg::Imu>("/imu/data_raw", 10);
        pub_gps_ = this->create_publisher<sensor_msgs::msg::NavSatFix>("/gps/fix", 10);
        pub_temp_ = this->create_publisher<sensor_msgs::msg::Temperature>("/imu/temperature", 10);
        pub_gps_vel_ = this->create_publisher<geometry_msgs::msg::TwistWithCovarianceStamped>("/gps/fix_velocity", 10);

        // CONCURRENCY: Background receiver worker
        running_ = true;
        rx_thread_ = std::thread(&ESP32Bridge::rx_loop, this);
    }

    ~ESP32Bridge() override {
        running_ = false;
        if (rx_thread_.joinable()) {
            rx_thread_.join();
        }
        uart_.close_port();
    }

private:
    UartDriver uart_;
    ProtocolParser parser_;
    std::atomic<bool> running_;
    std::thread rx_thread_;

    double effective_width_;
    std::string imu_frame_id_;
    std::string gps_frame_id_;

    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr sub_cmd_vel_;
    rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr pub_imu_;
    rclcpp::Publisher<sensor_msgs::msg::NavSatFix>::SharedPtr pub_gps_;
    rclcpp::Publisher<sensor_msgs::msg::Temperature>::SharedPtr pub_temp_;
    rclcpp::Publisher<geometry_msgs::msg::TwistWithCovarianceStamped>::SharedPtr pub_gps_vel_;

    // KINEMATICS: Converts Twist to skid-steer wheel speeds with 0.6 m/s clamp
    void cmd_vel_callback(const geometry_msgs::msg::Twist::SharedPtr msg) {
        double vx = msg->linear.x;
        double wz = msg->angular.z;

        double v_left = vx - (wz * effective_width_ / 2.0);
        double v_right = vx + (wz * effective_width_ / 2.0);

        // ROBOTICS: Proportional velocity scaling to preserve steering radius
        constexpr double MAX_SPEED = 0.6;
        double largest = std::max(std::abs(v_left), std::abs(v_right));
        if (largest > MAX_SPEED) {
            v_left = (v_left / largest) * MAX_SPEED;
            v_right = (v_right / largest) * MAX_SPEED;
        }

        CmdVelPayload payload{static_cast<float>(v_left), static_cast<float>(v_right)};
        send_packet(PKT_ID_CMD_VEL, reinterpret_cast<const uint8_t *>(&payload), sizeof(payload));
    }

    // PROTOCOL: Packages payload into protocol frame and sends via UART
    void send_packet(uint8_t pkt_id, const uint8_t *payload, uint8_t len) {
        uint8_t crc = ProtocolParser::calculate_crc(pkt_id, len, payload);

        std::vector<uint8_t> frame;
        frame.reserve(len + 5);
        frame.push_back(FRAME_HEADER1);
        frame.push_back(FRAME_HEADER2);
        frame.push_back(pkt_id);
        frame.push_back(len);
        frame.insert(frame.end(), payload, payload + len);
        frame.push_back(crc);

        uart_.write_bytes(frame.data(), frame.size());
    }

    // WORKER: Dedicated serial reception loop
    void rx_loop() {
        uint8_t byte = 0;
        uint8_t pkt_id = 0;
        uint8_t payload_len = 0;
        uint8_t payload[64];

        while (running_ && rclcpp::ok()) {
            ssize_t n = uart_.read_byte(byte);
            if (n <= 0) continue;

            if (parser_.parse_byte(byte, pkt_id, payload, payload_len)) {
                handle_incoming_packet(pkt_id, payload, payload_len);
            }
        }
    }

    // TELEMETRY: Routes unpacked payload to target ROS publisher
    void handle_incoming_packet(uint8_t pkt_id, const uint8_t *payload, uint8_t len) {
        auto stamp = this->now();

        if (pkt_id == PKT_ID_IMU && len == sizeof(ImuPayload)) {
            ImuPayload data;
            std::memcpy(&data, payload, sizeof(ImuPayload));

            auto msg = sensor_msgs::msg::Imu();
            msg.header.stamp = stamp;
            msg.header.frame_id = imu_frame_id_;
            msg.linear_acceleration.x = data.ax;
            msg.linear_acceleration.y = data.ay;
            msg.linear_acceleration.z = data.az;
            msg.angular_velocity.x = data.gx;
            msg.angular_velocity.y = data.gy;
            msg.angular_velocity.z = data.gz;
            msg.orientation_covariance[0] = -1.0;
            pub_imu_->publish(msg);

            // Sıcaklık yayını:
            auto temp_msg = sensor_msgs::msg::Temperature();
            temp_msg.header.stamp = stamp;
            temp_msg.header.frame_id = imu_frame_id_;
            temp_msg.temperature = data.temp;
            temp_msg.variance = 0.0;
            pub_temp_->publish(temp_msg);

        } else if (pkt_id == PKT_ID_GPS && len == sizeof(GpsPayload)) {
            GpsPayload data;
            std::memcpy(&data, payload, sizeof(GpsPayload));

            auto msg = sensor_msgs::msg::NavSatFix();
            msg.header.stamp = stamp;
            msg.header.frame_id = gps_frame_id_;

            // Fix Durumu
            if (data.fix > 0) {
                msg.status.status = sensor_msgs::msg::NavSatStatus::STATUS_FIX;
            } else {
                msg.status.status = sensor_msgs::msg::NavSatStatus::STATUS_NO_FIX;
            }
            msg.status.service = sensor_msgs::msg::NavSatStatus::SERVICE_GPS;

            msg.latitude = data.lat;
            msg.longitude = data.lon;
            msg.altitude = static_cast<double>(data.alt);

            // Kovaryans Hesabı (HDOP üzerinden yaklaşık doğruluk matrisi)
            if (data.fix > 0 && data.hdop > 0.0f) {
                // HDOP * tipik GPS standart sapması (~2.5m)
                double accuracy = data.hdop * 2.5;
                double variance = accuracy * accuracy;

                msg.position_covariance[0] = variance;      // Doğu-Batı
                msg.position_covariance[4] = variance;      // Kuzey-Güney
                msg.position_covariance[8] = variance * 4.0; // İrtifa (genelde 2x hata payı)
                msg.position_covariance_type = sensor_msgs::msg::NavSatFix::COVARIANCE_TYPE_APPROXIMATED;
            } else {
                msg.position_covariance[0] = 0.0;
                msg.position_covariance_type = sensor_msgs::msg::NavSatFix::COVARIANCE_TYPE_UNKNOWN;
            }

            pub_gps_->publish(msg);

            // GPS Hız Yayını
            if (data.fix > 0) {
                auto vel_msg = geometry_msgs::msg::TwistWithCovarianceStamped();
                vel_msg.header.stamp = stamp;
                vel_msg.header.frame_id = "base_link"; // Robotun ileri ekseni
                vel_msg.twist.twist.linear.x = static_cast<double>(data.speed); // m/s
                
                // Hız varyansı (GPS hız hassasiyeti genelde ~0.1 m/s'dir)
                vel_msg.twist.covariance[0] = 0.05; // vx varyansı
                pub_gps_vel_->publish(vel_msg);
            }
        }
    }
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ESP32Bridge>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}